src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs

```diff
@@ -283,6 +294,107 @@ protected override void ComputeDependencyNodeDependencies(List<DependencyNodeCor
             }
         }
 
+        private string ResilientNameComputation(MethodDesc method)
+        {
+            try
+            {
+                return method.ToString();
```

Comment:
I'll leave it up to you (I haven't actually tried how it would look like), but I think there might be value in updating the `MethodDesc.ToString` and `DebugNameFormatter` (that is used to generate the string for type `ToString()`) to be resilient like this, instead of having two algorithms. Three reasons:

* `ToString()` is used when debugging, so when actually debugging the compilation failure, we could enjoy usable names, instead of funceval just telling us that `ToString` crapped out. ToString is not supposed to throw.
* We would ensure the formatting is consistent (e.g. `ToString()` shortens System.Private to S.P.)
* We won't have 100 lines of very rarely exercised code that might have bugs (the amount of bugs in this code review worries me).

Yes, it's a bit more involved since this formatter is also used in the runtime type system, but I _think_ we should be able to achieve a pretty good effect with dot files (add a `DefType.Diag.cs` that add `DiagnosticName` and `DiagnosticNamespace` virtual properties - the runtime type system already has those - these would try to get the name but if it doesn't work out, could return tokens - and the same for methods).

We would also need to be more deliberate in where we use try/catch due to DebugNameFormatter's structure, but that's probably a good thing.

Reply:
Its certainly worth the effort to try to make it fit in there. I was worried it might be a rather larger change than I wanted in the single-exe prototype, but I won't claim that I like this logic as written at all, and I am proposing putting it into master... give me some time and I'll try to make a better approach.